packages/evm/src/opcodes/functions.ts

```diff
@@ -984,6 +1011,346 @@ export const handlers: Map<number, OpHandler> = new Map([
+    function (runState, _common) {
+      if (runState.env.eof === undefined) {
+        // Opcode not available in legacy contracts
+        trap(ERROR.INVALID_OPCODE)
+      }
+      const toExchange = Number(
+        bytesToBigInt(runState.code.subarray(runState.programCounter, runState.programCounter + 1))
+      )
+      const n = (toExchange >> 4) + 1
+      const m = (toExchange & 0x0f) + 1
+      runState.stack.exchange(n, n + m)
+      runState.programCounter++
+    },
+  ],
+  // 0xec: EOFCREATE
+  [
+    0xec,
+    async function (runState, _common) {
+      if (runState.env.eof === undefined) {
+        // Opcode not available in legacy contracts
+        trap(ERROR.INVALID_OPCODE)
+      } else {
+        // Read container index
+        const containerIndex = runState.env.code[runState.programCounter]
+        const containerCode = runState.env.eof!.container.body.containerSections[containerIndex]
+
+        // Pop stack values
+        const [value, salt, inputOffset, inputSize] = runState.stack.popN(4)
+
+        const gasLimit = runState.messageGasLimit!
+        runState.messageGasLimit = undefined
+
+        let data = new Uint8Array(0)
+        if (inputSize !== BIGINT_0) {
+          data = runState.memory.read(Number(inputOffset), Number(inputSize), true)
+        }
+
+        runState.programCounter++ // Jump over the immediate byte
+
+        const ret = await runState.interpreter.eofcreate(
+          gasLimit,
+          value,
+          containerCode,
+          setLengthLeft(bigIntToBytes(salt), 32),
+          data
+        )
+        runState.stack.push(ret)
+      }
+    },
+  ],
+  // 0xee: RETURNCONTRACT
+  [
+    0xee,
+    async function (runState, _common) {
+      if (runState.env.eof === undefined) {
+        // Opcode not available in legacy contracts
+        trap(ERROR.INVALID_OPCODE)
+      } else {
+        // Read container index
+        const containerIndex = runState.env.code[runState.programCounter]
+        const containerCode = runState.env.eof!.container.body.containerSections[containerIndex]
+
+        // Read deployContainer as EOFCreate (initcode) container
+        const deployContainer = new EOFContainer(containerCode, EOFContainerMode.Initmode)
+
+        // Pop stack values
+        const [auxDataOffset, auxDataSize] = runState.stack.popN(2)
+
+        let auxData = new Uint8Array(0)
+        if (auxDataSize !== BIGINT_0) {
+          auxData = runState.memory.read(Number(auxDataOffset), Number(auxDataSize))
+        }
+
+        const originalDataSize = deployContainer.header.dataSize
+        const preDeployDataSectionSize = deployContainer.body.dataSection.length
+        const actualSectionSize = preDeployDataSectionSize + Number(auxDataSize)
+
+        if (actualSectionSize < originalDataSize) {
+          trap(ERROR.INVALID_RETURNCONTRACT_DATA_SIZE)
+        }
+
+        if (actualSectionSize > 0xffff) {
+          // Data section size is now larger than the max data section size
```

Comment:
Could this ever happen in EOF?  I thought the code validation would prevent this?

Reply:
This can happen in the data section. The data section of a not-yet-deployed contract can be lower than, or equal to, the data size in the contract header. When the contract is being deployed, the RETURNCONTRACT will actually point to a memory region of data to append. 

I don't know why I added "Temp: trap OOG?" there, maybe this is not in the spec (but it still seems to pass the reference tests). So, you are right, this is actually one of the cases where an invalid container might be deployed.

Reply:
Ah, we can likely remove this. If you try to deploy a data section of size `0xffff`, you try to deploy a contract with higher byte length than the maximum, so it will go OOG anyways.